lightningd/invoice.c

```diff
@@ -1441,3 +1475,162 @@ static const struct json_command decodepay_command = {
 	"Decode {bolt11}, using {description} if necessary"
 };
 AUTODATA(json_command, &decodepay_command);
+
+/* If we fail because it exists, we also return the clashing invoice */
+static struct command_result *fail_exists(struct command *cmd,
+					  const struct json_escape *label)
+{
+	struct json_stream *data;
+	struct invoice invoice;
+	struct wallet *wallet = cmd->ld->wallet;
+
+	data = json_stream_fail(cmd, INVOICE_LABEL_ALREADY_EXISTS,
+				"Duplicate label");
+	if (!wallet_invoice_find_by_label(wallet, &invoice, label))
+		fatal("Duplicate invoice %s not found any more?",
+		      label->s);
+
+	json_add_invoice(data, wallet_invoice_details(cmd, wallet, invoice));
+	json_object_end(data);
+
+	return command_failed(cmd, data);
+}
+
+static struct command_result *json_createinvoice(struct command *cmd,
+						 const char *buffer,
+						 const jsmntok_t *obj UNNEEDED,
+						 const jsmntok_t *params)
+{
+	const char *invstring;
+	struct json_escape *label;
+	struct preimage *preimage;
+	struct invoice invoice;
+	struct sha256 payment_hash;
+	struct json_stream *response;
+	struct bolt11 *b11;
+	struct sha256 hash;
+	u5 *sig;
+	bool have_n;
+	char *fail;
+
+	if (!param(cmd, buffer, params,
+		   p_req("invstring", param_string, &invstring),
+		   p_req("preimage", param_preimage, &preimage),
+		   p_req("label", param_label, &label),
```

Comment:
If we make `preimage` optional, then `label` needs to move up in order to have all required params be first.

Reply:
Thanks for reordering, I'll try to tackle the `p_opt("preimage", ...)` next ^^